subprojects/core/src/main/java/org/gradle/initialization/definition/DefaultInjectedPluginDependencies.java

```diff
@@ -16,22 +16,67 @@
 
 package org.gradle.initialization.definition;
 
+import com.google.common.collect.Lists;
+import org.gradle.api.Transformer;
 import org.gradle.api.initialization.definition.InjectedPluginDependencies;
 import org.gradle.api.initialization.definition.InjectedPluginDependency;
+import org.gradle.api.internal.initialization.ClassLoaderScope;
 import org.gradle.plugin.management.internal.DefaultPluginRequests;
+import org.gradle.plugin.management.internal.InvalidPluginRequestException;
+import org.gradle.plugin.management.internal.PluginRequestInternal;
 import org.gradle.plugin.management.internal.PluginRequests;
+import org.gradle.plugin.use.PluginId;
+import org.gradle.util.CollectionUtils;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+
+import static org.gradle.util.CollectionUtils.collect;
 
 public class DefaultInjectedPluginDependencies implements InjectedPluginDependencies {
+    private final List<DefaultInjectedPluginDependency> dependencies = Lists.newArrayList();
+    private final ClassLoaderScope classLoaderScope;
+
+    public DefaultInjectedPluginDependencies(ClassLoaderScope classLoaderScope) {
+        this.classLoaderScope = classLoaderScope;
+    }
+
     @Override
     public InjectedPluginDependency id(String id) {
-        // TODO: Record plugins requested
-        return new DefaultInjectedPluginDependency(id);
+        DefaultInjectedPluginDependency injectedPluginDependency = new DefaultInjectedPluginDependency(id);
+        dependencies.add(injectedPluginDependency);
+        return injectedPluginDependency;
     }
 
-    /**
-     * TODO: Capture context (classloader scope) and turn these into SelfResolvingPluginRequest
-     */
     public PluginRequests getRequests() {
-        return DefaultPluginRequests.EMPTY;
+        if (dependencies.isEmpty()) {
+            return DefaultPluginRequests.EMPTY;
+        }
+        return new DefaultPluginRequests(listPluginRequests());
+    }
+
+    List<PluginRequestInternal> listPluginRequests() {
+        List<PluginRequestInternal> pluginRequests = collect(dependencies, new Transformer<PluginRequestInternal, DefaultInjectedPluginDependency>() {
+            public PluginRequestInternal transform(DefaultInjectedPluginDependency original) {
+                return new SelfResolvingPluginRequest(original.getId(), classLoaderScope);
+            }
+        });
+
+        Map<PluginId, Collection<PluginRequestInternal>> groupedById = CollectionUtils.groupBy(pluginRequests, new Transformer<PluginId, PluginRequestInternal>() {
+            public PluginId transform(PluginRequestInternal pluginRequest) {
+                return pluginRequest.getId();
+            }
+        });
+
+        // Check for duplicates
+        for (PluginId key : groupedById.keySet()) {
```

Comment:
Do we care? We don't fail if you add the same dependency the multiple times or apply the same plugin multiple times.

Reply:
I stole this from https://github.com/gradle/gradle/blob/31d85b581062c3576b656559273714321d451145/subprojects/core/src/main/java/org/gradle/plugin/use/internal/PluginRequestCollector.java#L98 

In the regular `plugins {}` block, we fail if there are duplicates.  Should we remove that behavior or change it? I can open an issue.

Reply:
We should remove it I think. We should deal with this kind of thing in resolution, where we can see all of the inputs and determine whether there are conflicts.

Reply:
OK, done https://github.com/gradle/gradle/issues/4193